packages/preferences/src/browser/preferences-tree-widget.ts

```diff
@@ -0,0 +1,352 @@
+
+    constructor(protected readonly widgetManager: WidgetManager,
+                protected readonly shell: ApplicationShell,
+                protected readonly preferenceService: PreferenceService,
+                protected readonly messageService: MessageService,
+                protected readonly userPreferenceProvider: UserPreferenceProvider,
+                protected readonly workspacePreferenceProvider: WorkspacePreferenceProvider) {
+        super();
+
+        this.id = PREFERENCES_CONTAINER_WIDGET_ID;
+        this.title.label = 'Preferences';
+        this.title.closable = true;
+        this.title.iconClass = 'fa fa-sliders';
+
+        this.editors = [];
+
+        this.onDirtyChangedEmitter = new Emitter<void>();
+        this.onDirtyChanged = this.onDirtyChangedEmitter.event;
+        this.save = () => {
+            this.editors.forEach(editor => editor.saveable.save());
+        };
+
+        this.toDispose.push(this.onDirtyChangedEmitter);
+    }
+
+    dispose(): void {
+        if (this.isDisposed) {
+            return;
+        }
+        super.dispose();
+        this.toDispose.dispose();
+    }
+
+    getEditorWidget(): EditorWidget {
+        return this.currentEditor;
+    }
+
+    protected async onAfterAttach(msg: Message) {
+        this.treeWidget = await this.widgetManager.getOrCreateWidget(PREFERENCES_TREE_WIDGET_ID) as TreeWidget;
+        (this.treeWidget as PreferencesTreeWidget).onPreferenceSelected(value => {
+            const preferenceName = Object.keys(value)[0];
+            const preferenceValue = value[preferenceName];
+            if (this.dirty) {
+                this.messageService.warn('Preferences editor(s) has/have unsaved changes');
+            } else {
+                this.preferenceService.set(preferenceName,
+                    preferenceValue,
+                    this.currentEditor.title.label === 'User Preferences'
+                        ? PreferenceScope.User
+                        : PreferenceScope.Workspace);
+            }
+        });
+
+        const editorsContainer = new PreferencesEditorsContainer(this.widgetManager, this.userPreferenceProvider, this.workspacePreferenceProvider);
+        editorsContainer.onInit(() => {
+            toArray(editorsContainer.widgets()).forEach(editor => {
+                const editorWidget = editor as EditorWidget;
+                this.editors.push(editorWidget);
+                const savable = editorWidget.saveable;
+                savable.onDirtyChanged(() => {
+                    this.onDirtyChangedEmitter.fire(undefined);
+                });
+            });
+        });
+        editorsContainer.onEditorChanged(editor => {
+            if (this.currentEditor) {
+                this.currentEditor.saveable.save();
+            }
+            this.currentEditor = editor;
+            const currentWidget = this.shell.currentWidget;
+            this.shell.currentChanged.emit({
```

Comment:
One can only listen to current and active widget changes, only the shell can emit. We should hide this API by replacing signals with events.

Reply:
Agreed. @vinokurig Why do we need to set the inner editor as current, here?

Reply:
`ApplicationShell` has a focus tracker that fires editor-changed events, but it didn't work if editor changed inside a widget, so need to fire the event manually.

Reply:
I created a PR on your fork https://github.com/vinokurig/theia/pull/1
@akosyakov Could you have alook, too?